tests/integration/test_upgrade.py

```diff
@@ -14,51 +14,47 @@
 logger = logging.getLogger(__name__)
 
 
-@pytest.mark.abort_on_fail
 @pytest.mark.usefixtures("deploy_current_stable")
-class TestUpgrade:
-    """A wrapper class for upgrade tests"""
-
-    async def test_upgrade(self, ops_test: OpsTest):
-        """Test upgrading from the current stable release works as expected."""
-
-        logger.info("Getting model status")
-        status = await ops_test.model.get_status()  # noqa: F821
-        logger.info(f"Status: {status}")
-        assert ops_test.model.applications[APP_NAME].status == ACTIVE_STATUS
-
-        address = status["applications"][APP_NAME]["units"][f"{APP_NAME}/0"]["address"]
-        url = f"http://{address}:8080/debug/status"
-        logger.info("Querying app address: %s", url)
-        r = requests.get(url, timeout=2.0)
-        assert r.status_code == 200
-        logger.info(f"Output = {r.json()}")
-
-        # Deploy the locally built charm and wait for active/idle status
-        logger.info("refreshing running application with the new local charm")
-
-        charm = await fetch_charm(ops_test)
-        await ops_test.model.applications[APP_NAME].refresh(
-            path=charm,
-            resources=get_charm_resources(),
+async def test_upgrade(ops_test: OpsTest):
+    """Test upgrading from the current stable release works as expected."""
+
+    logger.info("Getting model status")
+    status = await ops_test.model.get_status()  # noqa: F821
+    logger.info(f"Status: {status}")
+    assert ops_test.model.applications[APP_NAME].status == ACTIVE_STATUS
+
+    address = status["applications"][APP_NAME]["units"][f"{APP_NAME}/0"]["address"]
+    url = f"http://{address}:8080/debug/status"
+    logger.info("Querying app address: %s", url)
+    r = requests.get(url, timeout=2.0)
+    assert r.status_code == 200
+    logger.info(f"Output = {r.json()}")
+
+    # Deploy the locally built charm and wait for active/idle status
+    logger.info("refreshing running application with the new local charm")
+
+    charm = await fetch_charm(ops_test)
+    await ops_test.model.applications[APP_NAME].refresh(
+        path=charm,
+        resources=get_charm_resources(),
+    )
+
+    logger.info("waiting for the upgraded unit to be ready")
+    async with ops_test.fast_forward():
+        await ops_test.model.wait_for_idle(
```

Comment:
I wouldn't use wait_for_idle with fast_forward, see my conversation here https://matrix.to/#/!xdClnUGkurzjxqiQcN:ubuntu.com/$Dgg6oqWbCPtuTqEDnzt_evWlScoM7jggBnFl-9TofMI?via=ubuntu.com&via=matrix.org
But it might be specific only to traefik-k8s

Reply:
Done.